src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java

```diff
@@ -197,23 +253,71 @@ String getTopicNameFromNamespace(final String prefix, final BsonDocument namespa
                 cfg.getString(DATABASE_CONFIG), cfg.getString(COLLECTION_CONFIG)));
     }
 
+    /**
+     * Checks to see if data should be copied.
+     *
+     * <p>
+     * Copying data is only required if it's been configured and it hasn't already completed.
+     * This method also is responsible for caching the {@code resumeAfter} value for the change stream.
+     * </p>
+     *
+     * @return true if should copy the existing data.
+     */
+    private boolean shouldCopyData() {
+        Map<String, Object> offset = getOffset(sourceConfig);
+        if (sourceConfig.getBoolean(COPY_EXISTING_CONFIG) && (offset == null || offset.containsKey("copy"))) {
+            MongoChangeStreamCursor<ChangeStreamDocument<Document>> changeStreamCursor =
+                    getChangeStreamIterable(sourceConfig, mongoClient).cursor();
+            ChangeStreamDocument<Document> firstResult = changeStreamCursor.tryNext();
+            if (firstResult != null) {
+                cachedResult = new BsonDocumentWrapper<>(firstResult, ChangeStreamDocument.createCodec(Document.class,
+                        MongoClientSettings.getDefaultCodecRegistry()));
+            }
+            cachedResumeAfter = firstResult != null ? firstResult.getResumeToken() : changeStreamCursor.getResumeToken();
+            changeStreamCursor.close();
+            return true;
+        }
+        return false;
+    }
+
     /**
      * Returns the next document to be delivered to Kafka.
      *
      * <p>
-     * Return the next result from the change stream cursor. Creating the cursor if necessary.
+     * <ol>
+     * <li>If copying data is in progress, returns the next result.</li>
+     * <li>If copying data and all data has been copied and there is a cached result return the cached result.</li>
+     * <li>Otherwise, return the next result from the change stream cursor. Creating a new cursor if necessary.</li>
+     * </ol>
+     *
      * </p>
      *
      * @return the next document
      */
     private Optional<BsonDocument> getNextDocument() {
+        if (isCopying.get()) {
+            Optional<BsonDocument> result = copyDataManager.poll();
+            if (result.isPresent() || copyDataManager.isCopying()) {
+                return result;
+            }
+
+            // No longer copying
+            isCopying.set(false);
+            if (cachedResult != null) {
+                result = Optional.of(cachedResult);
```

Comment:
Is there a test of this beyond ensuring that the count of documents is correct?

Reply:
We test that that documents from the copying manager look correct and we test the correct number of items are on a topic.